client/block-builder/src/lib.rs

```diff
@@ -136,44 +140,54 @@ where
 						extrinsics.push(xt);
 						Ok(())
 					}
-					Err(tx_validity) => Err(ApplyExtrinsicFailed::Validity(tx_validity).into())?,
+					Err(tx_validity) => Err(ApplyExtrinsicFailed::Validity(tx_validity).into()),
 				}
 			})
 		}
 	}
 
-	/// Consume the builder to return a valid `Block` containing all pushed extrinsics.
-	pub fn bake(mut self) -> Result<Block, ApiErrorFor<A, Block>> {
-		self.bake_impl()?;
-		Ok(<Block as BlockT>::new(self.header, self.extrinsics))
-	}
-
-	fn bake_impl(&mut self) -> Result<(), ApiErrorFor<A, Block>> {
-		self.header = self.api.finalize_block_with_context(
+	/// Consume the builder to "bake" a valid `Block` containing all pushed extrinsics.
+	///
+	/// Returns the build `Block`, the changes to the storage and an optional `StorageProof`.
+	/// The storage proof will be `Some(_)` when proof recording was enabled.
+	pub fn bake(mut self) -> Result<
+		(
+			Block,
+			StorageChanges<backend::StateBackendFor<B, Block>, Block>,
+			Option<StorageProof>
+		),
+		ApiErrorFor<A, Block>
+	> {
+		let header = self.api.finalize_block_with_context(
 			&self.block_id, ExecutionContext::BlockConstruction
 		)?;
 
 		debug_assert_eq!(
-			self.header.extrinsics_root().clone(),
+			header.extrinsics_root().clone(),
 			HashFor::<Block>::ordered_trie_root(
 				self.extrinsics.iter().map(Encode::encode).collect(),
 			),
 		);
 
-		Ok(())
-	}
+		let proof = self.api.extract_proof();
 
-	/// Consume the builder to return a valid `Block` containing all pushed extrinsics
-	/// and the generated proof.
-	///
-	/// The proof will be `Some(_)`, if proof recording was enabled while creating
-	/// the block builder.
-	pub fn bake_and_extract_proof(mut self)
-		-> Result<(Block, Option<StorageProof>), ApiErrorFor<A, Block>>
-	{
-		self.bake_impl()?;
+		let state = self.backend.state_at(self.block_id)?;
+		let changes_trie_storage = self.backend.changes_trie_storage();
+		let parent_hash = self.parent_hash;
 
-		let proof = self.api.extract_proof();
-		Ok((<Block as BlockT>::new(self.header, self.extrinsics), proof))
+		// The unsafe is required because the consume requires that we drop/consume the inner api
```

Comment:
If we have this https://github.com/rust-lang/rfcs/blob/master/text/1598-generic_associated_types.md it would be solvable without `unsafe`. I also have some ideas to try with the current implementation of rust, however that was really I point where I stopped and just took the "short" path.

Reply:
(The code in this function works without unsafe code, I just marked the function unsafe because there is this invariant that needs to be fulfilled.)

Reply:
If it is the invariant that needs to be fullfilled, should we file an issue for this, maybe even record your ideas, how to resolve that, so we don't forget that we want to and/or have other people spend time on making this cleaner (if the solution is actually cleaner...)?